python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,180 @@
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self.__class__._instances[(self._ptr, self._size)]) > 1
+
+    @property
+    def ptr(self) -> int:
+        """Device pointer to the start of the buffer."""
+        self._unlink_shared_buffers(zero_copied=True)
+        return self._ptr
+
+    @property
+    def mutable_ptr(self) -> int:
+        """Device pointer to the start of the buffer."""
+        self._unlink_shared_buffers()
+        return self._ptr
+
+    def _getitem(self, offset: int, size: int) -> Buffer:
+        """
+        Sub-classes can overwrite this to implement __getitem__
+        without having to handle non-slice inputs.
+        """
+        return self._from_device_memory(
+            cuda_array_interface_wrapper(
+                ptr=self._ptr + offset, size=size, owner=self.owner
+            )
+        )
+
+    def copy(self, deep: bool = True):
+        """
+        Return a copy of Buffer.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            If True, returns a deep-copy of the underlying Buffer data.
+            If False, returns a shallow-copy of the Buffer pointing to
+            the same underlying data.
+
+        Returns
+        -------
+        Buffer
+        """
+        if not deep and not self._zero_copied:
+            copied_buf = CopyOnWriteBuffer.__new__(CopyOnWriteBuffer)
+            copied_buf._ptr = self._ptr
+            copied_buf._size = self._size
+            copied_buf._owner = self._owner
+            copied_buf._finalize_init()
+            return copied_buf
+        else:
+            return self._from_device_memory(
+                rmm.DeviceBuffer(ptr=self._ptr, size=self.size)
+            )
+
+    @property
+    def __cuda_array_interface__(self) -> dict:
+        # Unlink if there are any weak-references.
+
+        # Mark the Buffer as ``zero_copied=True``,
+        # which will prevent any copy-on-write
+        # mechanism post this operation.
+        # This is done because we don't have any
+        # control over knowing if a third-party library
+        # has modified the data this Buffer is
+        # pointing to.
+        self._unlink_shared_buffers(zero_copied=True)
```

Comment:
`_unlink_shared_buffers` really seems to be doing two things: (a) "unlinking" `self` from any other buffers, and (b) setting `self.zero_copied=True`. Can we make it so that it only does one thing? We can explicitly set `self.zero_copied = True` as a separate statement.

Reply:
Separated these two steps.